### PR TITLE
feat: Add thirdpartypasswordless recipe

### PR DESCRIPTION
**Pull request by nkshah2**
## Summary
- Adds thirdpartypasswordless recipe
- Adds UserType export to thirdparty

## Remaining TODOs
- [ ] Verify size limit

### CODE REVIEW

File: lib/ts/recipe/thirdpartypasswordless/types.ts [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+    getStateAndOtherInfoFromStorage: <CustomStateProperties>(input: {
+        userContext: any;
+    }) => (StateObject & CustomStateProperties) | undefined;
+
+    setStateAndOtherInfoToStorage: <CustomStateProperties>(input: {
+        state: StateObject & CustomStateProperties;
+        userContext: any;
+    }) => void;
+
+    getAuthorizationURLWithQueryParamsAndSetState: (input: {
```
Comment (rishabhpoddar) [inline lib/ts/recipe/thirdpartypasswordless/types.ts:88]: This is called `Authorization`, but the other function is `Authorisation`. We need to make them consistent everywhere?

File: lib/ts/recipe/thirdpartypasswordless/types.ts [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+        providerId: string;
+        authorisationURL: string;
+        userContext: any;
+        providerClientId?: string;
+        options?: RecipeFunctionOptions;
+    }) => Promise<string>;
+
+    generateStateToSendToOAuthProvider: (input: { userContext: any }) => string;
+
+    verifyAndGetStateOrThrowError: <CustomStateProperties>(input: {
```
Comment (rishabhpoddar) [inline lib/ts/recipe/thirdpartypasswordless/types.ts:98]:
```suggestion
    verifyAndGetThirdPartyStateOrThrowError: <CustomStateProperties>(input: {
```

File: lib/ts/recipe/thirdpartypasswordless/types.ts [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+        stateFromAuthProvider: string | undefined;
+        stateObjectFromStorage: (StateObject & CustomStateProperties) | undefined;
+        userContext: any;
+    }) => Promise<StateObject & CustomStateProperties>;
+
+    getAuthCodeFromURL: (input: { userContext: any }) => string;
```
Comment (rishabhpoddar) [inline lib/ts/recipe/thirdpartypasswordless/types.ts:104]:
```suggestion
    getThirdPartyAuthCodeFromURL: (input: { userContext: any }) => string;
```

File: lib/ts/recipe/thirdpartypasswordless/types.ts [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+
+    getAuthErrorFromURL: (input: { userContext: any }) => string | undefined;
```
Comment (rishabhpoddar) [inline lib/ts/recipe/thirdpartypasswordless/types.ts:106]:
```suggestion
    getThirdPartyAuthErrorFromURL: (input: { userContext: any }) => string | undefined;
```

File: lib/ts/recipe/thirdpartypasswordless/types.ts [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+
+    getAuthStateFromURL: (input: { userContext: any }) => string;
```
Comment (rishabhpoddar) [inline lib/ts/recipe/thirdpartypasswordless/types.ts:108]:
```suggestion
    getThirdPartyAuthStateFromURL: (input: { userContext: any }) => string;
```

File: lib/ts/recipe/thirdpartypasswordless/types.ts [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+
+    createCode: (
```
Comment (rishabhpoddar) [inline lib/ts/recipe/thirdpartypasswordless/types.ts:110]:
```suggestion
    createPasswordlessCode: (
```

File: lib/ts/recipe/thirdpartypasswordless/types.ts [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+            | { phoneNumber: string; userContext: any; options?: RecipeFunctionOptions }
+    ) => Promise<{
+        status: "OK";
+        deviceId: string;
+        preAuthSessionId: string;
+        flowType: PasswordlessFlowType;
+        fetchResponse: Response;
+    }>;
+
+    resendCode: (input: {
```
Comment (rishabhpoddar) [inline lib/ts/recipe/thirdpartypasswordless/types.ts:122]:
```suggestion
    resendPasswordlessCode: (input: {
```

File: lib/ts/recipe/thirdpartypasswordless/types.ts [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+        userContext: any;
+        deviceId: string;
+        preAuthSessionId: string;
+        options?: RecipeFunctionOptions;
+    }) => Promise<{
+        status: "OK" | "RESTART_FLOW_ERROR";
+        fetchResponse: Response;
+    }>;
+
+    consumeCode: (
```
Comment (rishabhpoddar) [inline lib/ts/recipe/thirdpartypasswordless/types.ts:132]:
```suggestion
    consumePasswordlessCode: (
```

File: lib/ts/recipe/thirdpartypasswordless/recipe.ts [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+                          postAPIHook: config.postAPIHook,
+                          override: {
+                              emailVerification: config.override?.emailVerification,
+                              functions: function () {
+                                  return DerivedThirdPartyRecipeImplementation(_recipeImplementation);
+                              },
+                          },
+                      },
+                      {
+                          emailVerification: this.emailVerificationRecipe,
```
Comment (rishabhpoddar) [inline lib/ts/recipe/thirdpartypasswordless/recipe.ts:75]: Where is `this.emailVerificationRecipe`? How are the tests passing?

File: lib/ts/recipe/passwordless/utils.ts [hunk 10]
```diff
@@ -27,3 +29,146 @@ export function normaliseUserInput(config: InputType): NormalisedInputType {
         override,
     };
 }
+
```
Comment (rishabhpoddar) [inline lib/ts/recipe/passwordless/utils.ts:32]: Add a comment as to why these functions exist..

File: lib/ts/recipe/thirdpartypasswordless/index.ts [hunk 11]
```diff
@@ -0,0 +1,230 @@
...
+              fetchResponse: Response;
+          }
+    > {
+        return Recipe.getInstanceOrThrow().recipeImplementation.thirdPartySignInAndUp({
+            ...input,
+            userContext: getNormalisedUserContext(input?.userContext),
+        });
+    }
+
+    static getAuthorizationURLWithQueryParamsAndSetState(input: {
```
Comment (rishabhpoddar) [inline lib/ts/recipe/thirdpartypasswordless/index.ts:54]:
```suggestion
    static getThirdPartyAuthorizationURLWithQueryParamsAndSetState(input: {
```

File: lib/ts/recipe/thirdpartypasswordless/index.ts [hunk 12]
```diff
@@ -0,0 +1,230 @@
...
+    }> {
+        const recipe: Recipe = Recipe.getInstanceOrThrow();
+
+        return PasswordlessUtilsFunctions.resendCode({
+            ...input,
+            recipeImplementation: recipe.passwordlessRecipe.recipeImplementation,
+        });
+    }
+
+    static async consumeCode(
```
Comment (rishabhpoddar) [inline lib/ts/recipe/thirdpartypasswordless/index.ts:98]:
```suggestion
    static async consumePasswordlessCode(
```

File: lib/ts/recipe/thirdpartypasswordless/index.ts [hunk 13]
```diff
@@ -0,0 +1,230 @@
...
+    }> {
+        const recipe: Recipe = Recipe.getInstanceOrThrow();
+
+        return PasswordlessUtilsFunctions.createCode({
+            ...input,
+            recipeImplementation: recipe.passwordlessRecipe.recipeImplementation,
+        });
+    }
+
+    static async resendCode(input: { userContext?: any; options?: RecipeFunctionOptions }): Promise<{
```
Comment (rishabhpoddar) [inline lib/ts/recipe/thirdpartypasswordless/index.ts:86]:
```suggestion
    static async resendPasswordlessCode(input: { userContext?: any; options?: RecipeFunctionOptions }): Promise<{
```

File: lib/ts/recipe/thirdpartypasswordless/index.ts [hunk 14]
```diff
@@ -0,0 +1,230 @@
...
+        providerClientId?: string;
+        options?: RecipeFunctionOptions;
+    }): Promise<string> {
+        return Recipe.getInstanceOrThrow().recipeImplementation.getAuthorizationURLWithQueryParamsAndSetState({
+            ...input,
+            userContext: getNormalisedUserContext(input.userContext),
+        });
+    }
+
+    static async createCode(
```
Comment (rishabhpoddar) [inline lib/ts/recipe/thirdpartypasswordless/index.ts:67]:
```suggestion
    static async createPasswordlessCode(
```

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: lib/ts/recipe/thirdpartypasswordless/recipe.ts (hunk 9 above)
Comment (nkshah2) [reply lib/ts/recipe/thirdpartypasswordless/recipe.ts:75]: The recipe class extends AuthRecipeWithEmailVerification, `this.emailVerificationRecipe` comes from the super class

Comment (rishabhpoddar) [reply lib/ts/recipe/thirdpartypasswordless/recipe.ts:75]: I see. Ok

File: lib/ts/recipe/thirdpartypasswordless/types.ts (hunk 1 above)
Comment (nkshah2) [reply lib/ts/recipe/thirdpartypasswordless/types.ts]: Changed this everywhere to use `Authorisation`

**Review by rishabhpoddar — APPROVED** [review]

**Pull request merged**